src/main/kotlin/org/rust/lang/core/resolve/LangItems.kt

```diff
@@ -35,13 +36,39 @@ fun findIteratorItemType(project: Project, ty: RustType): RustType {
     return rawType.substitute(typeParameterMap)
 }
 
+fun findIndexOutputType(project: Project, containerType: RustType, indexExprType: RustType): RustType {
+    val impls = RsImplIndex.findImpls(project, containerType)
+    for (impl in impls) {
+        val trait = impl.traitRef?.resolveToTrait ?: continue
+        if (trait.isIndex) {
+            // TODO: get index type from impl declaration
+            val indexType = impl.functionList
+                .find { it.name == "index" }
+                ?.valueParameterList
+                ?.valueParameterList
+                // 'index' function have only one value parameter
+                // fn index(&self, index: Idx) -> &Self::Output;
+                ?.getOrNull(0)
+                ?.typeReference
+                ?.type ?: continue
+            if (indexType.canUnifyWith(indexExprType, project)) {
```

Comment:
Let's make the check more optimistic: if there's only impl `Index` for the type (common case), we should just use it, without the check that `indexExprType` match.